### PR TITLE
Added a ton of missing tests, added the facility to CaseDTO and cleaned up some technical debt.

### PR DESCRIPTION
**Pull request by AndreSoderlund**

### CODE REVIEW

File: src/test/java/se/sundsvall/casemanagement/integration/casedata/CaseDataServiceTest.java [hunk 1]
```diff
@@ -241,6 +240,17 @@ void testGetStatusNotFound() {
 			.hasFieldOrPropertyWithValue("detail", Constants.ERR_MSG_STATUS_NOT_FOUND);
 	}
 
+	@Test
+	void patchErrandWithAttachmentNotFound() {
+		final var id = UUID.randomUUID().toString();
+		when(caseDataClientMock.postAttachment(any())).thenThrow(Problem.valueOf(Status.NOT_FOUND));
+
+		assertThatThrownBy(() -> caseDataService.patchErrandWithAttachment(id, List.of(new AttachmentDTO())))
```
Comment (cheezi0747) [inline src/test/java/se/sundsvall/casemanagement/integration/casedata/CaseDataServiceTest.java:248]: Refactor the code of the lambda to have only one invocation possibly throwing a runtime exception.

File: src/test/java/se/sundsvall/casemanagement/integration/casedata/CaseDataServiceTest.java [hunk 2]
```diff
@@ -241,6 +240,17 @@ void testGetStatusNotFound() {
 			.hasFieldOrPropertyWithValue("detail", Constants.ERR_MSG_STATUS_NOT_FOUND);
 	}
 
+	@Test
+	void patchErrandWithAttachmentNotFound() {
+		final var id = UUID.randomUUID().toString();
+		when(caseDataClientMock.postAttachment(any())).thenThrow(Problem.valueOf(Status.NOT_FOUND));
```
Comment (cheezi0747) [inline src/test/java/se/sundsvall/casemanagement/integration/casedata/CaseDataServiceTest.java:246]: be more specific than any()

File: src/main/java/se/sundsvall/casemanagement/api/model/ByggRCaseDTO.java [hunk 3]
```diff
@@ -6,24 +6,24 @@
...
-@EqualsAndHashCode(callSuper = true)
-@ToString(callSuper = true)
 @Data
-public class PlanningPermissionCaseDTO extends CaseDTO implements Serializable {
+@ToString(callSuper = true)
+@EqualsAndHashCode(callSuper = true)
+@Schema(description = "ByggR-cases")
+public class ByggRCaseDTO extends CaseDTO implements Serializable {
+
+	private String diaryNumber;
```
Comment (cheezi0747) [inline src/main/java/se/sundsvall/casemanagement/api/model/ByggRCaseDTO.java:22]: Schema annotations

File: src/main/java/se/sundsvall/casemanagement/api/model/AttachmentDTO.java [hunk 4]
```diff
@@ -4,12 +4,18 @@
...
 import io.swagger.v3.oas.annotations.media.Schema;
+import lombok.AllArgsConstructor;
+import lombok.Builder;
 import lombok.Data;
+import lombok.NoArgsConstructor;
 
 @Data
+@Builder(setterPrefix = "with")
+@NoArgsConstructor
+@AllArgsConstructor
```
Comment (cheezi0747) [inline src/main/java/se/sundsvall/casemanagement/api/model/AttachmentDTO.java:18]: Schema annotations

File: src/main/java/se/sundsvall/casemanagement/api/model/FacilityDTO.java [hunk 5]
```diff
@@ -1,27 +1,49 @@
...
+import lombok.Setter;
+
+
+@Getter
+@Setter
+@Builder(setterPrefix = "with")
+@NoArgsConstructor
+@AllArgsConstructor
+@JsonPropertyOrder({"facilityType", "description", "address", "facilityCollectionName", "mainFacility", "extraParameters"})
+public class FacilityDTO {
```
Comment (cheezi0747) [inline src/main/java/se/sundsvall/casemanagement/api/model/FacilityDTO.java:27]: Schema annotations

File: src/main/java/se/sundsvall/casemanagement/api/model/OtherCaseDTO.java [hunk 6]
```diff
@@ -11,4 +12,6 @@
 @Data
 public class OtherCaseDTO extends CaseDTO implements Serializable {
```
Comment (cheezi0747) [inline src/main/java/se/sundsvall/casemanagement/api/model/OtherCaseDTO.java:13]: Schema annotations

File: src/main/java/se/sundsvall/casemanagement/api/model/PersonDTO.java
Comment (cheezi0747) [inline src/main/java/se/sundsvall/casemanagement/api/model/PersonDTO.java:1]: Schema annotations

File: src/main/java/se/sundsvall/casemanagement/integration/casedata/CaseDataMapper.java [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+
+	private static final String APPLICATION_PRIORITY_KEY = "application.priority";
+
+	private CaseDataMapper() {
+		// Prevent instantiation
+	}
+
+	public static generated.client.casedata.AttachmentDTO toAttachment(final AttachmentDTO attachment, final String errandNumber) {
+		final generated.client.casedata.AttachmentDTO attachmentDTO = new generated.client.casedata.AttachmentDTO();
+		attachmentDTO.setCategory(attachment.getCategory());
```
Comment (cheezi0747) [inline src/main/java/se/sundsvall/casemanagement/integration/casedata/CaseDataMapper.java:34]: use builder pattern? goes for the entire class

File: src/main/java/se/sundsvall/casemanagement/integration/ecos/PartyService.java [hunk 8]
```diff
@@ -162,9 +162,10 @@ private ArrayOfguid getEcosFacilityRoles(final StakeholderDTO s) {
 			.withGuid(s.getRoles().stream()
 				.map(roleString ->
 				{
-					final var r	= StakeholderRole.valueOf(roleString);
+					final var r = StakeholderRole.valueOf(roleString);
```
Comment (cheezi0747) [inline src/main/java/se/sundsvall/casemanagement/integration/ecos/PartyService.java:165]: final var role perhaps? (and the parameter could be something cooler than s perhaps :D

File: src/test/java/se/sundsvall/casemanagement/TestUtil.java [hunk 9]
```diff
@@ -236,6 +184,20 @@ public static StakeholderDTO createStakeholder(final StakeholderType stakeholder
 		}
 	}
 
+	public static OtherCaseDTO createOtherCaseDTO() {
+		final var otherCase = new OtherCaseDTO();
```
Comment (cheezi0747) [inline src/test/java/se/sundsvall/casemanagement/TestUtil.java:188]: builder pattern letsgoooo

File: src/test/java/se/sundsvall/casemanagement/TestUtil.java [hunk 10]
```diff
@@ -147,38 +98,38 @@ public static AttachmentDTO createAttachmentDTO(final AttachmentCategory attachm
...
+					getHeatPumpExtraParams();
+				default -> createExtraParameters();
+			});
 		return facility;
 	}
 
-	public static PlanningPermissionCaseDTO createPlanningPermissionCaseDTO(final CaseType caseType, final AttachmentCategory attachmentCategory) {
-		final PlanningPermissionCaseDTO pCase = new PlanningPermissionCaseDTO();
+	public static ByggRCaseDTO createPlanningPermissionCaseDTO(final CaseType caseType, final AttachmentCategory attachmentCategory) {
+		final ByggRCaseDTO pCase = new ByggRCaseDTO();
```
Comment (cheezi0747) [inline src/test/java/se/sundsvall/casemanagement/TestUtil.java:121]: with the change to ByggrCaseDto pcase could be ByggCase or ByggRCase

File: src/test/java/se/sundsvall/casemanagement/api/AttachmentResourceTest.java [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+	private WebTestClient webTestClient;
+
+	@Test
+	void postAttachmentsToCase_Ecos() {
+		final var attachments = List.of(createAttachment(AttachmentCategory.BUILDING_PERMIT_APPLICATION));
+		final var caseMapping = createCaseMapping(c -> c.setSystem(SystemType.ECOS));
+
+		when(caseMappingService.getCaseMapping(PATH_VARIABLE)).thenReturn(caseMapping);
+
+		webTestClient.post().uri(uriBuilder -> uriBuilder.path("/cases/{externalCaseId}/attachments").build(PATH_VARIABLE))
```
Comment (cheezi0747) [inline src/test/java/se/sundsvall/casemanagement/api/AttachmentResourceTest.java:55]: uri on a new line....please.. 🥺

File: src/test/java/se/sundsvall/casemanagement/api/CaseResourceTest.java [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+			.exchange()
+			.expectStatus().isOk()
+			.expectBody(CaseResourceResponseDTO.class)
+			.returnResult()
+			.getResponseBody();
+
+		assertThat(result.getCaseId()).isEqualTo("Inskickat");
+
+		verify(caseService).handleCase(caseDTOCaptor.capture());
+		assertThat(caseDTOCaptor.getValue()).satisfies(c -> {
```
Comment (cheezi0747) [inline src/test/java/se/sundsvall/casemanagement/api/CaseResourceTest.java:72]: whats a c?

File: src/test/java/se/sundsvall/casemanagement/api/CaseStatusResourceTest.java [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+		when(caseMappingService.getCaseMapping("externalCaseId")).thenReturn(caseMapping);
+		when(ecosService.getStatus(caseMapping.getCaseId(), caseMapping.getExternalCaseId())).thenReturn(caseStatusDTO);
+
+		final var response = webTestClient.get().uri(uriBuilder -> uriBuilder.path("cases/{externalCaseId}/status").build("externalCaseId"))
+			.exchange()
+			.expectStatus().isOk()
+			.expectBody(CaseStatusDTO.class)
+			.returnResult().getResponseBody();
+
+		assertThat(response).satisfies(r -> {
```
Comment (cheezi0747) [inline src/test/java/se/sundsvall/casemanagement/api/CaseStatusResourceTest.java:101]: whats a r? this is java not some fancy pants language. be verbose

File: src/test/java/se/sundsvall/casemanagement/service/CaseServiceTest.java [hunk 14]
```diff
@@ -61,7 +61,7 @@ class CaseServiceTest {
 	@Test
 	void testHandleByggRCase() {
 		// Arrange
-		final var pCase = new PlanningPermissionCaseDTO();
+		final var pCase = new ByggRCaseDTO();
```
Comment (cheezi0747) [inline src/test/java/se/sundsvall/casemanagement/service/CaseServiceTest.java:64]: byggcase

File: src/test/java/se/sundsvall/casemanagement/integration/db/model/CaseEntityTest.java [hunk 15]
```diff
@@ -44,7 +44,7 @@ void testFields() throws SQLException {
 		// Arrange
 		final var id = "123";
 		final var deliveryStatus = DeliveryStatus.CREATED;
-		final var environmentalCaseDTO = new EnvironmentalCaseDTO();
+		final var environmentalCaseDTO = new EcosCaseDTO();
```
Comment (cheezi0747) [inline src/test/java/se/sundsvall/casemanagement/integration/db/model/CaseEntityTest.java:47]: ecoscase

File: src/main/java/se/sundsvall/casemanagement/api/validation/OneMainFacility.java
Comment (josc04) [inline src/main/java/se/sundsvall/casemanagement/api/validation/OneMainFacility.java:1]: OneMainFacility and OneMainFacilityOrNull could be merged into one (OneMainFacility) if validator is extended with the option to be nullable (see `@ValidUuid(nullable = true)` for example).

File: src/main/java/se/sundsvall/casemanagement/api/validation/impl/OneMainFacilityOrNullConstraintValidator.java [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+
+public class OneMainFacilityOrNullConstraintValidator implements ConstraintValidator<OneMainFacilityOrNull, List<FacilityDTO>> {
+
+	@Override
+	public boolean isValid(List<FacilityDTO> planningPermissionFacilityList, ConstraintValidatorContext context) {
+		if (planningPermissionFacilityList == null) {
+			// We can't do this validation if the list is null
+			return true;
+		}
+		return planningPermissionFacilityList.stream().filter(FacilityDTO::isMainFacility).count() <= 1;
```
Comment (josc04) [inline src/main/java/se/sundsvall/casemanagement/api/validation/impl/OneMainFacilityOrNullConstraintValidator.java:19]: The validator is named `OneMainFacilityOrNullConstraintValidator`. Shouldn't count be equal to 1 here for the input to be valid? If no facility is main facility this validatror will return true, is that correct?

If the validator is verifying that no more than one facility is main, maybe rename validator to MaxOneMainFacilityValidator?

File: src/test/java/se/sundsvall/casemanagement/api/CaseResourceFailureTest.java [hunk 17]
```diff
@@ -0,0 +1,194 @@
...
+import se.sundsvall.casemanagement.api.model.CaseResourceResponseDTO;
+import se.sundsvall.dept44.test.annotation.resource.Load;
+import se.sundsvall.dept44.test.extension.ResourceLoaderExtension;
+
+@SpringBootTest(classes = Application.class, webEnvironment = RANDOM_PORT)
+@ActiveProfiles("junit")
+@ExtendWith(ResourceLoaderExtension.class)
+class CaseResourceFailureTest {
+
+	@Autowired
```
Comment (josc04) [inline src/test/java/se/sundsvall/casemanagement/api/CaseResourceFailureTest.java:35]: No mockbeans for service-classes and verification of them not being used?

File: src/test/java/se/sundsvall/casemanagement/api/validation/ByggRStakeholderRoleConstraintValidatorTest.java [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+import se.sundsvall.casemanagement.api.validation.impl.ByggRStakeholderRoleConstraintValidator;
+
+@ExtendWith(MockitoExtension.class)
+class ByggRStakeholderRoleConstraintValidatorTest {
```
Comment (josc04) [inline src/test/java/se/sundsvall/casemanagement/api/validation/ByggRStakeholderRoleConstraintValidatorTest.java:18]: Add test for empty string and null as value?

File: src/main/java/se/sundsvall/casemanagement/api/validation/impl/ByggRStakeholderRoleConstraintValidator.java [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+public class ByggRStakeholderRoleConstraintValidator implements ConstraintValidator<ByggRStakeholderRole, List<String>> {
+
+	@Override
+	public boolean isValid(List<String> roles, ConstraintValidatorContext context) {
+		return roles.stream().allMatch(this::isValidRole);
+	}
+
+	private boolean isValidRole(String role) {
+
+		if (role == null || role.isEmpty()) {
```
Comment (josc04) [inline src/main/java/se/sundsvall/casemanagement/api/validation/impl/ByggRStakeholderRoleConstraintValidator.java:21]: isEmpty check is redundant here as it will be invalid when checked against the list with valid roles at row 34.

File: src/main/java/se/sundsvall/casemanagement/api/validation/impl/ByggRStakeholderRoleConstraintValidator.java [hunk 20]
```diff
@@ -0,0 +1,36 @@
...
+			return false;
+		}
+
+		// Valid roles
+		final Set<String> validRoles = Set.of(
```
Comment (josc04) [inline src/main/java/se/sundsvall/casemanagement/api/validation/impl/ByggRStakeholderRoleConstraintValidator.java:26]: This can be moved to be a static list instead of creating it every time method is called.

**Review by josc04 — APPROVED** [review]
